ssl/statem/statem_srvr.c

```diff
@@ -101,13 +101,28 @@ int ossl_statem_server_read_transition(SSL *s, int mt)
          *      b) We are running SSL3 (in TLS1.0+ the client must return a 0
          *         list if we requested a certificate)
          */
-        if (mt == SSL3_MT_CLIENT_KEY_EXCHANGE
-                && (!s->s3->tmp.cert_request
-                    || (!((s->verify_mode & SSL_VERIFY_PEER) &&
-                          (s->verify_mode & SSL_VERIFY_FAIL_IF_NO_PEER_CERT))
-                        && (s->version == SSL3_VERSION)))) {
-            st->hand_state = TLS_ST_SR_KEY_EXCH;
-            return 1;
+        if (mt == SSL3_MT_CLIENT_KEY_EXCHANGE) {
+            if (s->s3->tmp.cert_request) {
+                if (s->version == SSL3_VERSION) {
+                    if ((s->verify_mode & SSL_VERIFY_PEER) &&
```

Comment:
It makes no sense for SSL_VERIFY_FAIL_IF_NO_PEER_CERT to be set without SSL_VERIFY_PEER. The documentation says you must set both if you want SSL_VERIFY_FAIL_IF_NO_PEER_CERT. SSL_CTX_set_verify() doesn't enforce that though.

I couldn't see a bad indent? I changed to break before "&&"